### PR TITLE
add IsRemote field to SpanContext, set by propagators

### LINKED ISSUE #187: How to know a parent is remote when starting a span
Whether or not the parent of a span that is being started is remote is an attribute that must be passed to samplers.

[Span creation](https://github.com/open-telemetry/opentelemetry-specification/blob/0d255935cc6569a6f195daaaf361453dbd9cb5ac/specification/tracing-api.md#span-creation) docs do mention remote parents but it isn't clear how it is deciding it is remote:

>  tracer.StartSpan(options.WithNoParent(), options.WithParentContext(parentCtx)) will generate a new child span with remote parent parentCtx, and WithNoParent will be ignored.

Why is it assumed that `parentCtx` is remote here?

In OpenCensus we had a optional remote parent flag that could be set when calling `start_span`.

### PR DESCRIPTION
**Pull request by tsloughter**
Without a field to mark if a `SpanContext` came from a remote process `Span` creation and sampling can not know if the new span has a remote parent.

This PR adds a `IsRemote` field to `SpanContext` and defines that it must be set when deserializing a propagated `SpanContext`.

This is additionally important for implementations that can not pass the parent `Span` as an argument to `startSpan` but only a `SpanContext`. That is the case with the Erlang/Elixir implementations because the content of a `Span` is kept in a table (since it needs to be able to be mutated) while the `SpanContext` is an immutable record.

Closes #187

### CODE REVIEW

**Comment by bogdandrutu** [comment]
@tsloughter let's wait couple of days until we put the RFC for sampling, the hasRemoteParent may be removed, otherwise I really like this proposal if the remote parent stays.

**Comment by tsloughter** [comment]
@bogdandrutu ok, sounds good. Curious to see the sampling RFC.

**Comment by bogdandrutu** [comment]
Please see the RFC proposal https://github.com/open-telemetry/rfcs/pull/24

**Comment by tsloughter** [comment]
Now that the sampling rfc has been merged is this ready?

**Comment by yurishkuro** [comment]
it is indeed the property of the context itself. E.g. in this OpenTracing example:
```
ctx = tracer.extract(Format.HTTP, request.headers)
span = tracer.startSpan('name', childOf=ctx)
```

the `ctx` here is "remote".

File: specification/api-sampling.md [hunk 1]
```diff
@@ -30,7 +30,7 @@ Returns the sampling Decision for a `Span` to be created.
 - `SpanContext` of a parent `Span`. Typically extracted from the wire. Can be
   `null`.
 - Boolean that indicates that `SpanContext` was extracted from the wire, i.e.
-  parent `Span` is from the different process.
+  parent `Span` is from the different process, meaning `IsRemote` is set to true.
```
Comment (Oberon00) [inline specification/api-sampling.md]: If that boolean is contained in the SpanContext, I'd remove the extra argument altogether.

**Comment by Oberon00** [comment]
What about, instead of a boolean, the SpanContext having an optional reference to the in-process Span, `null` meaning that it is remote?

File: specification/api-sampling.md (hunk 1 above)
Comment (tsloughter) [reply specification/api-sampling.md]: Good point, probably can remove the boolean argument.

**Comment by tsloughter** [comment]
@Oberon00 that is an interesting idea since in the Erlang impl I currently do similar. The term stored when a span is created is `{SpanCtx, Parent}` where `Parent` is another `{SpanCtx, Parent}` and so on. This allows for easily making the parent the current span context when finishing a span.

**Comment by bogdandrutu** [comment]
@Oberon00 Having a reference to the "Span" causes a circular dependencies because Span also has a reference to the SpanContext. I think we should not do that.

**Comment by Oberon00** [comment]
>  Span also has a reference to the SpanContext

Not necessarily. It could just store the required info directly (or in some SpanContextData object that is equivalent to SpanContext minus back-reference) and create a SpanContext on demand.

Also, I don't think a circular dependency is really a problem here, Span and SpanContext are closely related anyway.

**Comment by yurishkuro** [comment]
@tsloughter do you mind elaborating why you need isRemote? What would startSpan do differently?

**Comment by tsloughter** [comment]
@yurishkuro it is used in the probability sampler https://github.com/open-telemetry/oteps/blob/eeab6a2d1342abb58ad5e7c0aadaaa4b1ab32310/text/0006-sampling.md#default-samplers

```
Allows users to configure if probability applies only for "root spans", "root spans and remote parent", or "all spans".

    Default is to apply only for "root spans and remote parent".
    Remote parent property should be added to the SpanContext see specs PR/216
```

**Comment by yurishkuro** [comment]
I would suggest putting the circular reference discussion aside, as it is not related to this issue, only as an implementation detail. If we want to support configurable upsamping scenarios at "local root" spans, and expose it via sampling API (not just implementation), then we need isRemote in the API.

**Pull request merged**